File: frontend/src/engine/capabilities/addComment.ts
/**
 * addComment – Add a comment/note to a cell.
 *
 * Uses the Office.js Comments API (ExcelApi 1.10+).
 * Falls back to adding a note if comments API is unavailable.
 */

import { CapabilityMeta, AddCommentParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange } from "./rangeUtils";

const meta: CapabilityMeta = {
  action: "addComment",
  description: "Add a comment or note to a cell",
  mutates: true,
  affectsFormatting: false,
  requiresApiSet: "ExcelApi 1.10",
};

async function handler(
  context: Excel.RequestContext,
  params: AddCommentParams,
  options: ExecutionOptions
): Promise<StepResult> {
  if (options.dryRun) {
    return { stepId: "", status: "success", message: `Would add comment to ${params.cell}` };
  }

  options.onProgress?.("Adding comment...");

  const range = resolveRange(context, params.cell);
  context.workbook.comments.add(range, params.content);
  await context.sync();

  return {
    stepId: "",
    status: "success",
    message: `Added comment to ${params.cell}: "${params.content.slice(0, 50)}${params.content.length > 50 ? "…" : ""}"`,
  };
}

// ── Legacy-Excel fallback (ExcelApi < 1.10) ──────────────────────────────────
// workbook.comments.add requires 1.10. We emulate a cell comment by writing
// the text into the adjacent right-hand cell with italic light-yellow
// styling, prefixed with "Note:". Fidelity cost: no hover popup, no author/
// timestamp, no threaded replies, and the note consumes a real cell (the
// handler checks that the adjacent cell is empty before overwriting). For
// most annotation intents this is close enough.
async function fallback(
  context: Excel.RequestContext,
  params: AddCommentParams,
  options: ExecutionOptions,
): Promise<StepResult> {
  if (options.dryRun) {
    return {
      stepId: "",
      status: "success",
      message: `Would emit inline note next to ${params.cell} (legacy fallback).`,
    };
  }

  options.onProgress?.("Legacy-Excel mode: writing inline note (comments unavailable)...");

  const anchor = resolveRange(context, params.cell);
  // Move one column to the right. Office.js Range.getOffsetRange(rowOffset, colOffset).
  const noteCell = anchor.getOffsetRange(0, 1);
  noteCell.load(["values", "address"]);
  await context.sync();

  const existing = (noteCell.values?.[0]?.[0] ?? "") as string | number | boolean | null;
  if (existing !== null && existing !== "" && existing !== undefined) {
    // Don't overwrite real user data. Bail loudly but not as a hard error.
    return {
      stepId: "",
      status: "success",
      message:
        `Could not write inline note: the cell adjacent to ${params.cell} is occupied. ` +
        `Comments require ExcelApi 1.10+; add the note manually (legacy-Excel fallback).`,
    };
  }

  const preview = params.content.length > 50 ? params.content.slice(0, 50) + "…" : params.content;
  noteCell.values = [[`Note: ${params.content}`]];
  noteCell.format.font.italic = true;
  noteCell.format.font.color = "#705000";
  noteCell.format.fill.color = "#FFF8D6"; // light yellow
  noteCell.format.wrapText = true;

  await context.sync();

  return {
    stepId: "",
    status: "success",
    message:
      `Wrote inline note next to ${params.cell}: "${preview}" — native comments require ` +
      `ExcelApi 1.10+ (legacy-Excel fallback; consumes one cell, no hover popup).`,
  };
}

registry.register(meta, handler as any, fallback as any);
export { meta };

File: backend/app/execution/capabilities/add_comment.py
"""addComment — attach a comment to a cell."""

from __future__ import annotations

from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    address = params.get("cell") or params.get("range")
    text = params.get("comment") or params.get("text")
    if not address or not text:
        return {"status": "error", "message": "addComment requires 'cell' and 'comment'."}

    rng = resolve_range(ctx.workbook_handle, address)
    if ctx.dry_run:
        return {"status": "preview", "message": f"Would add comment to {rng.address}."}

    try:
        # Delete any existing comment first for idempotency.
        if rng.api.Comment is not None:
            rng.api.Comment.Delete()
        rng.api.AddComment(str(text))
    except Exception as exc:  # noqa: BLE001
        return {"status": "error", "message": f"Add comment failed: {exc}", "error": str(exc)}

    return {
        "status": "success",
        "message": f"Added comment to {rng.address}.",
        "outputs": {"cell": rng.address},
    }


registry.register("addComment", handler, mutates=True)
